File: 2-D Permutations.py
"""
TLE, CHECK C++ CODE

https://dmoj.ca/problem/2dperm
typical annoying ad hoc question

Using offline queries (definitely not intended solution)
Note: the intended solution uses a difference array (and doesn't have that extra log factor from sorting)

TC: O(MN*log(MN) + Qlog(Q))
"""

import sys

input = sys.stdin.readline
N, M, Q = map(int, input().split())
area = N * M
table = sorted(i * j for i in range(1, N + 1) for j in range(1, M + 1))

queries = [(int(input()), i) for i in range(Q)]
queries.sort(key=lambda x: x[0], reverse=True)  # sort by value
ans = [0] * Q

i = 0
j = area - 1
for num, idx in queries:
    while i + 1 < area and table[i + 1] + num - 1 <= area:
        i += 1
    while j - 1 >= 0 and table[j - 1] > num:
        j -= 1

    ans[idx] = i - (area - j) + 1
    if num == area:  # weird corner case
        ans[idx] = 1

print("\n".join(map(str, ans)))

# # 50/100, brute force code
# N, M, Q = map(int, input().split())
# area = N * M
#
# for _ in range(Q):
#     q = int(input())
#     total = 0
#
#     for i in range(1, N + 1):
#         for j in range(1, M + 1):
#             less = i * j
#             more = (N-i+1) * (M-j+1)
#             if less + q - 1 <= area and more <= q:
#                 total += 1
#
#     print(total)

File: 2-D Permutations.cpp
/*
 https://dmoj.ca/problem/2dperm
 typical annoying ad hoc question

 Using offline queries (definitely not intended solution)
 Note: the intended solution uses a difference array (and doesn't have that extra log factor from sorting)

 TC: O(MN*log(MN) + Qlog(Q))
*/

#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    int N, M, Q;
    cin >> N >> M >> Q;
    int area = N * M;
    vector<int> table;
    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= M; j++) {
            table.push_back(i * j);
        }
    }
    sort(table.begin(), table.end());

    vector<pair<int, int>> queries(Q);
    for (int i = 0; i < Q; i++) {
        cin >> queries[i].first;
        queries[i].second = i;
    }
    sort(queries.rbegin(), queries.rend());

    vector<int> ans(Q);
    int i = 0, j = area - 1;
    for (auto [num, idx] : queries) {
        while (i + 1 < area && table[i + 1] + num - 1 <= area) {
            i++;
        }
        while (j - 1 >= 0 && table[j - 1] > num) {
            j--;
        }

        ans[idx] = i - (area - j) + 1;
        if (num == area) {
            ans[idx] = 1;
        }
    }

    for (int i = 0; i < Q; i++) {
        cout << ans[i] << "\n";
    }

    return 0;
}
